### DISCUSSION TITLE
2019/07/03/linux-tuning

### DISCUSSION THREAD

**Discussion by giscus** [wanghenshui/wanghenshui.github.io#111] [Q&A]
# 2019/07/03/linux-tuning

https://wanghenshui.github.io/2019/07/03/linux-tuning.html

**Comment by wanghenshui** [comment]
backlog 参数描述的是服务器端 TCP ESTABELLISHED 状态对应的全连接队列长度。

全连接队列长度如何计算？
如果 backlog 大于内核参数 net.core.somaxconn，则以 net.core.somaxconn 为准，
即全连接队列长度 = min(backlog, 内核参数 net.core.somaxconn)，net.core.somaxconn 默认为 128。
这个很好理解，net.core.somaxconn 定义了系统级别的全连接队列最大长度，
backlog 只是应用层传入的参数，不可能超过内核参数，所以 backlog 必须小于等于 net.core.somaxconn。


半连接队列长度如何计算？
半连接队列长度由内核参数 tcp_max_syn_backlog 决定，
当使用 SYN Cookie 时（就是内核参数 net.ipv4.tcp_syncookies = 1），这个参数无效，
半连接队列的最大长度为 backlog、内核参数 net.core.somaxconn、内核参数 tcp_max_syn_backlog 的最小值。
即半连接队列长度 = min(backlog, 内核参数 net.core.somaxconn，内核参数 tcp_max_syn_backlog)。
这个公式实际上规定半连接队列长度不能超过全连接队列长度。

https://jaminzhang.github.io/linux/understand-Linux-backlog-and-somaxconn-kernel-arguments/

**Comment by wanghenshui** [comment]
对于高并发的短连接应用来说，在没有及时调用accept进行处理时，会导致accept queue队列满。

在队列满的情况下，内核参数/proc/sys/net/ipv4/tcp_abort_on_overflow默认是0，操作系统不会给连接回复RST，而是默默的将ACK丢弃。会导致server对第二个包SYN+ACK进行重传，连接也无法建立。

https://xiking.win/2020/04/21/redis-request-time-cost-troubleshoot/

**Reply by wanghenshui** [reply]
https://xiking.win/2020/05/18/2-linux-kernel-net-packet-dropped/

**Reply by wanghenshui** [reply]
https://xiking.win/2020/04/21/redis-request-time-cost-troubleshoot/ 这个是POD CPU限制导致回包慢，bug
